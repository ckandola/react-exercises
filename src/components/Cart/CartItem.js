import React from 'react';
import { useGlobalContext } from './index';
import PropTypes from 'prop-types';

const CartItem = ({id, img, title, price, amount}) => {
    const { removeCartItem, increaseCartAmt, decreaseCartAmt } = useGlobalContext();
    return (
        <article className="cart-item">
            <img src={img} alt={title}/>
            <div>
                <h4>{title}</h4>
                <h4 className="cart-item-price">
                    ${price}
                </h4>
                <button 
                    className="cart-remove-btn"
                    onClick={() => removeCartItem(id)}>
                    Remove
                </button>
            </div>
            <div>
                <button 
                    className="cart-amount-btn"
                    onClick={() => increaseCartAmt(id)}>
                    <svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 20 20'>
                        <path d='M10.707 7.05L10 6.343 4.343 12l1.414 1.414L10 9.172l4.243 4.242L15.657 12z' />
                    </svg>
                </button>
                <p className="cart-amount">{amount}</p>
                <button 
                    className="cart-amount-btn"
                    onClick={() => decreaseCartAmt(id)}>
                    <svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 20 20'>
                        <path d='M9.293 12.95l.707.707L15.657 8l-1.414-1.414L10 10.828 5.757 6.586 4.343 8z' />
                    </svg>
                </button>
            </div>
        </article>
    );
};

export default CartItem;

// id and price from the tutorial link will be strings; data.js has them as numbers.

CartItem.propTypes = {
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    img: PropTypes.string,
    title: PropTypes.string,
    price: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    amount: PropTypes.number
};
