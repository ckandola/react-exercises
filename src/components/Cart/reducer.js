const reducer = (state, action) => {
    switch(action.type) {
        case 'CLEAR_CART':
            return {...state, cart:[]};
        default:
            console.log(`unknown action: ${action.type}`);
            break;
    }
    return state;
}

export default reducer;
