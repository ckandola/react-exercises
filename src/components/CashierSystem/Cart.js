import React from 'react';
import { FaCashRegister } from 'react-icons/fa';
import { useGlobalContext } from './index';

const Cart = () => {
    const { posCartState } = useGlobalContext();

    return (
        <section className="pos-cart">
            {posCartState.map((item, index) => {
                return (
                    <div key={index}>{item.id}</div>
                )
            })}
            <FaCashRegister className="pos-cart-icon" />
        </section>
    );
};

export default Cart;
