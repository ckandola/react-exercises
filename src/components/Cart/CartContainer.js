import React from 'react';
import CartItem from './CartItem';
import { useGlobalContext } from './index';

const CartContainer = () => {
    const {cart} = useGlobalContext();

    return (
        <section className="cart">
            <header>
                <h2>Your Bag</h2>
            </header>

            {cart.length === 0 ?
                <h4>Is Currently Empty</h4>
                :
                <div>
                    <div>
                        {cart.map(cartItem => {
                            return <CartItem key={cartItem.id} {...cartItem} />
                        })}
                    </div>
                    <footer>
                        <hr className="cart-hr"/>
                        <div className="cart-total">
                            <h4>
                                Total <span>$0.00</span>
                            </h4>
                        </div>
                        <button className="cart-btn cart-clear-btn">
                            Clear Cart
                        </button>
                    </footer>
                </div>
            }
        </section>
    );
};

export default CartContainer;
