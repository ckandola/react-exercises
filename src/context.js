import React, { useState, useContext, useReducer, useEffect, useCallback } from 'react';
import {default as smenulinks} from './components/StripeMenu/data';

import cartItems from './components/Cart/data';
import {default as cartReducer} from './components/Cart/reducer';

const AppContext = React.createContext();
    // gives both Provider and Consumer

const AppProvider = ({children}) => {
    const [isModalOpen, setIsModalOpen] = useState(false);
    
    // StripeMenu ---

    const [isSidebarOpen, setIsSidebarOpen] = useState(false);
    const [isSubmenuOpen, setIsSubmenuOpen] = useState(false);
    const [location, setLocation] = useState({});
    const [page, setPage] = useState({page: '', links:[]});

    const openSidebar = () => {
        setIsSidebarOpen(true);
    }
    
    const closeSidebar = () => {
        setIsSidebarOpen(false);
    }
    
    const openSubmenu = (text, coordinates) => {
        setPage(smenulinks.find(x => x.page === text));
        setLocation(coordinates);
        setIsSubmenuOpen(true);
    }
    
    const closeSubmenu = () => {
        setIsSubmenuOpen(false);
    }

    // Modal ---

    const openModal = () => {
        setIsModalOpen(true);
    }

    const closeModal = () => {
        setIsModalOpen(false);
    }

    // Cart ---
    const cartUrl = 'https://course-api.com/react-useReducer-cart-project';
    const initialCartState = {
        cartLoading: false,
        cart: cartItems,
        cartTotal: 0,
        cartAmount: 0
    };
    const [cartState, cartDispatch] = useReducer(cartReducer, initialCartState);
    const clearCart = () => {
        cartDispatch({type: 'CLEAR_CART'});
    };
    const removeCartItem = id => {
        cartDispatch({type: 'REMOVE_ITEM', payload: id});
    }
    const increaseCartAmt = id => {
        cartDispatch({type: 'INCREASE', payload: id});
    }
    const decreaseCartAmt = id => {
        cartDispatch({type: 'DECREASE', payload: id});
    }
    const fetchCartData = async () => {
        cartDispatch({type: 'LOADING'});
        const response = await fetch(cartUrl);
        const cart = await response.json();
        cartDispatch({type: 'DISPLAY_ITEMS', payload: cart});
    }

    useEffect(() => {
        cartDispatch({type: 'GET_TOTALS'});
    }, [cartState.cart]);

    useEffect(() => {
        fetchCartData();
    }, []);

    // CocktailDB ---
    const cocktailUrl = 'https://www.thecocktaildb.com/api/json/v1/1/search.php?s=';
    const [cocktailLoading, setCocktailLoading] = useState(true);
    const [cocktailSearchTerm, setCocktailSearchTerm] = useState('a');
    const [cocktails, setCocktails] = useState([]);

    const fetchDrinks = useCallback(async () => {
        try {
            setCocktailLoading(true);
            const response = await fetch(`${cocktailUrl}${cocktailSearchTerm}`);
            const data = await response.json();
            const {drinks} = data;
            if (!drinks) {
                setCocktails([]);
            } else {
                const newDrinks = drinks.map(item => {
                    const {idDrink, strDrink, strDrinkThumb, strAlcoholic, strGlass} = item;
                    return {id: idDrink, name: strDrink, img: strDrinkThumb, isAlcoholic: strAlcoholic, glass: strGlass};
                });
                setCocktails(newDrinks);
            }
            setCocktailLoading(false);
        } catch (error) {
            console.error(error);
            setCocktailLoading(false);
        }
    }, [cocktailSearchTerm]);

    useEffect(() => {
        fetchDrinks();
    }, [fetchDrinks, cocktailSearchTerm]);

    return (
        <AppContext.Provider value={{
            isModalOpen, openModal, closeModal,
            
            isSidebarOpen, openSidebar, closeSidebar,
            isSubmenuOpen, openSubmenu, closeSubmenu,
            location, page,

            ...cartState, clearCart, removeCartItem, 
            increaseCartAmt, decreaseCartAmt, fetchCartData,

            cocktailLoading, cocktails, setCocktailSearchTerm
        }}>
            {children}
        </AppContext.Provider>
    );
};

// custom hook
export const useGlobalContext = () => {
    return useContext(AppContext);
};

export {AppContext, AppProvider};
