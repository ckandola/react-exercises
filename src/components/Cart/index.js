// from https://youtu.be/a_7Z7C_JCyo?t=24176

import React from 'react';
import { useGlobalContext } from '../../context';
import Navbar from './Navbar';
import CartContainer from './CartContainer';
import './Cart.css';

const Cart = () => {
    const { cartLoading } = useGlobalContext();
    if (cartLoading) {
        return (
            <div className="cart-loading">
                <h1>Loading...</h1>
            </div>
        );
    }
    return (
        <main>
            <Navbar />
            <CartContainer />
        </main>
    );
};

export { useGlobalContext };
export default Cart;
