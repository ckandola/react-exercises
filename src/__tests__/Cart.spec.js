import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Cart from '../components/Cart';
import * as AppProvider from '../context';
import reducer from '../components/Cart/reducer';

describe('Cart tests', () => {
    it('matches the snapshot', () => {
        const placeholderFn = () => {};
        const contextValues = {
            cartLoading: false, cartAmount: 0, cart: [], cartTotal: 0,
            clearCart: placeholderFn, removeCartItem: placeholderFn, 
            increaseCartAmt: placeholderFn, decreaseCartAmt: placeholderFn, 
            fetchCartData: placeholderFn
        };
        jest.spyOn(AppProvider, 'useGlobalContext').mockImplementation(() => contextValues);
        const wrapper = mount(<Cart />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it(`increments a cart item's quantity when its up arrow is clicked`, () => {
        const placeholderFn = () => {};
        const mockIncr = jest.fn();
        const contextValues = {
            cartLoading: false, cartAmount: 3, 
            cart: [{ id: 0, title: 'test2', price: 0.50, img: './temp.img', amount: 3}], 
            cartTotal: 1.50,
            clearCart: placeholderFn, removeCartItem: placeholderFn,
            increaseCartAmt: mockIncr, decreaseCartAmt: placeholderFn,
            fetchCartData: placeholderFn
        };
        jest.spyOn(AppProvider, 'useGlobalContext').mockImplementation(() => contextValues);
        const wrapper = mount(<Cart />);
        wrapper.find('.cart-amount-btn').at(0).simulate('click');
        expect(mockIncr.mock.calls.length).toBe(1);
    });

    it(`decrements a cart item's quantity when its down arrow is clicked`, () => {
        const placeholderFn = () => {};
        const mockDecr = jest.fn();
        const contextValues = {
            cartLoading: false, cartAmount: 4, 
            cart: [{ id: 0, title: 'test3', price: 0.50, img: './temp.img', amount: 4}], 
            cartTotal: 2.00,
            clearCart: placeholderFn, removeCartItem: placeholderFn,
            increaseCartAmt: placeholderFn, decreaseCartAmt: mockDecr,
            fetchCartData: placeholderFn
        };
        jest.spyOn(AppProvider, 'useGlobalContext').mockImplementation(() => contextValues);
        const wrapper = mount(<Cart />);
        wrapper.find('.cart-amount-btn').at(1).simulate('click');
        expect(mockDecr.mock.calls.length).toBe(1);
    });

    it(`removes an item from the cart entirely when its REMOVE button is clicked`, () => {
        const placeholderFn = () => {};
        const mockRem = jest.fn();
        const contextValues = {
            cartLoading: false, cartAmount: 4, 
            cart: [{ id: 0, title: 'test3', price: 0.50, img: './temp.img', amount: 4}], 
            cartTotal: 2.00,
            clearCart: placeholderFn, removeCartItem: mockRem,
            increaseCartAmt: placeholderFn, decreaseCartAmt: placeholderFn,
            fetchCartData: placeholderFn
        };
        jest.spyOn(AppProvider, 'useGlobalContext').mockImplementation(() => contextValues);
        const wrapper = mount(<Cart />);
        wrapper.find('.cart-remove-btn').simulate('click');
        expect(mockRem.mock.calls.length).toBe(1);
    });

    it(`clears the cart when the CLEAR CART button is clicked`, () => {
        const placeholderFn = () => {};
        const mockClear = jest.fn();
        const contextValues = {
            cartLoading: false, cartAmount: 4, 
            cart: [{ id: 0, title: 'test3', price: 0.50, img: './temp.img', amount: 4}], 
            cartTotal: 2.00,
            clearCart: mockClear, removeCartItem: placeholderFn,
            increaseCartAmt: placeholderFn, decreaseCartAmt: placeholderFn,
            fetchCartData: placeholderFn
        };
        jest.spyOn(AppProvider, 'useGlobalContext').mockImplementation(() => contextValues);
        const wrapper = mount(<Cart />);
        wrapper.find('.cart-clear-btn').simulate('click');
        expect(mockClear.mock.calls.length).toBe(1);
    });

    describe('reducer tests', () => {
        it('clears the cart when `action.type` is CLEAR_CART', () => {
            const state = {
                cartLoading: false,
                cart: [1, 2, 3, 4],
                cartTotal: 50,
                cartAmount: 4
            };
            expect(reducer(state, {type: 'CLEAR_CART'})).toEqual({cartLoading: false, cart: [], cartTotal: 50, cartAmount: 4});
        });

        it('removes the second item from the cart when `action.type` is REMOVE_ITEM and `action.payload` matches the id', () => {
            const state = {
                cartLoading: false,
                cart: [{id: 1}, {id: 2}, {id: 3}, {id: 4}],
                cartTotal: 50,
                cartAmount: 4
            };
            expect(reducer(state, {type: 'REMOVE_ITEM', payload: 2})).toEqual({
                cartLoading: false, cart: [{id: 1}, {id: 3}, {id: 4}], cartTotal: 50, cartAmount: 4
            });
        });

        it('increments quantity of an item by 1 when `action.type` is INCREASE and `action.payload` matches the id', () => {
            const state = {
                cartLoading: false,
                cart: [{id: 1, amount: 1}, {id: 2, amount: 7}, {id: 3, amount: 5}],
                cartTotal: 26,
                cartAmount: 13
            };
            expect(reducer(state, {type: 'INCREASE', payload: 3})).toEqual({
                cartLoading: false,
                cart: [{id: 1, amount: 1}, {id: 2, amount: 7}, {id: 3, amount: 6}],
                cartTotal: 26,
                cartAmount: 13
            });
        });

        it('decrements quantity of an item by 1 when `action.type` is DECREASE and `action.payload` matches the id', () => {
            const state = {
                cartLoading: false,
                cart: [{id: 1, amount: 1}, {id: 2, amount: 7}, {id: 3, amount: 6}],
                cartTotal: 26,
                cartAmount: 13
            };
            expect(reducer(state, {type: 'DECREASE', payload: 2})).toEqual({
                cartLoading: false,
                cart: [{id: 1, amount: 1}, {id: 2, amount: 6}, {id: 3, amount: 6}],
                cartTotal: 26,
                cartAmount: 13
            });
        });

        it('removes the item from the cart entirely when `action.type` is DECREASE and its quantity is 1', () => {
            const state = {
                cartLoading: false,
                cart: [{id: 1, amount: 1}, {id: 2, amount: 6}, {id: 3, amount: 6}],
                cartTotal: 26,
                cartAmount: 13
            };
            expect(reducer(state, {type: 'DECREASE', payload: 1})).toEqual({
                cartLoading: false,
                cart: [{id: 2, amount: 6}, {id: 3, amount: 6}],
                cartTotal: 26,
                cartAmount: 13
            });
        });

        it('returns accurate total and quantity when `action.type` is GET_TOTALS', () => {
            const state = {
                cartLoading: false,
                cart: [{id: 2, amount: 6, price: 1.55}, {id: 3, amount: 6, price: 7.00}],
                cartTotal: 26,
                cartAmount: 13
            };
            expect(reducer(state, {type: 'GET_TOTALS'})).toEqual({
                cartLoading: false,
                cart: [{id: 2, amount: 6, price: 1.55}, {id: 3, amount: 6, price: 7.00}],
                cartTotal: 51.30,
                cartAmount: 12
            });
        });

        it('sets `cartLoading` to `true` when `action.type` is LOADING', () => {
            const state = {
                cartLoading: false,
                cart: [{id: 2, amount: 6, price: 1.55}, {id: 3, amount: 6, price: 7.00}],
                cartTotal: 51.30,
                cartAmount: 12
            };
            expect(reducer(state, {type: 'LOADING'})).toEqual({
                cartLoading: true,
                cart: [{id: 2, amount: 6, price: 1.55}, {id: 3, amount: 6, price: 7.00}],
                cartTotal: 51.30,
                cartAmount: 12
            });
        });

        it('returns a new cart, not loading, if a new cart is provided as `action.payload` when `action.type` is DISPLAY_ITEMS', () => {
            const state = {
                cartLoading: true,
                cart: [{id: 2, amount: 6, price: 1.55}, {id: 3, amount: 6, price: 7.00}],
                cartTotal: 51.30,
                cartAmount: 12
            };
            const newCart = [{id: 1, amount: 2, price: 4.40}, {id: 19, amount: 2, price: 7.75}];
            expect(reducer(state, {type: 'DISPLAY_ITEMS', payload: newCart})).toEqual({
                cartLoading: false,
                cart: [{id: 1, amount: 2, price: 4.40}, {id: 19, amount: 2, price: 7.75}],
                cartTotal: 51.30,
                cartAmount: 12
            });
        });
    });

});
