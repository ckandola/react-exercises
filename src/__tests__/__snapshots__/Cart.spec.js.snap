// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart tests matches the snapshot 1`] = `
<Cart>
  <main>
    <Navbar>
      <nav
        className="cart-nav"
      >
        <div
          className="cart-nav-center"
        >
          <h3>
            useReducer
          </h3>
          <div
            className="cart-nav-container"
          >
            <svg
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M16 6v2h2l2 12H0L2 8h2V6a6 6 0 1 1 12 0zm-2 0a4 4 0 1 0-8 0v2h8V6zM4 10v2h2v-2H4zm10 0v2h2v-2h-2z"
              />
            </svg>
            <div
              className="cart-amount-container"
            >
              <p
                className="cart-total-amount"
              >
                0
              </p>
            </div>
          </div>
        </div>
      </nav>
    </Navbar>
    <CartContainer>
      <section
        className="cart"
      >
        <header>
          <h2>
            Your Bag
          </h2>
        </header>
        <header>
          <h4
            style={
              Object {
                "margin": "auto",
                "width": "28%",
              }
            }
          >
            Is Currently Empty
          </h4>
        </header>
      </section>
    </CartContainer>
  </main>
</Cart>
`;
